### ISSUE TITLE
The parsed results in not a tree even ensure_tree = True

### ISSUE THREAD

**Issue by iclementine** [jcyk/Dynet-Biaffine-dependency-parser#9]
I rad the code about the decoding process of the parser, mainly the `arc_argmax` in `lib/utils.py` and `tarjan.py`, but I thought there is only guarantees that:

1. only one element is the child of <ROOT>,
2. acyclic
3. single head for every node except for the <ROOT>.

But, there is no guarantee for the result to be connected. I have a piece of code for testing:

```
import numpy as np
import dynet as dy

from lib import arc_argmax

m = np.random.randn(10, 10)
mt = dy.inputTensor(m)
mtp = dy.softmax(mt, d=1)
probs = mtp.npvalue()

mask = np.array([1,1,1,1,1,1,1,1,0,0]) # so only 7 words is valid. if includes the <ROOT> , then 8
sent_len = np.sum(mask) # 8

heads = arc_argmax(probs, sent_len, mask)
dependents = range(1, sent_len)
for d, h in zip(dependents, heads[1: sent_len]):
	print('{0} --> {1}'.format(d, h))
```
Though I use randomly initialized m as the logits, but trying multiple times will produce some result like this.

```
1 --> 3
2 --> 5
3 --> 2
4 --> 1
5 --> 0
6 --> 5
7 --> 3
```
~~The graph is not connected.  What's the problem, did I make something wrong? Thank you!~~

Sorry, I made a mistake about the offset (<ROOT> is included).
